### PR TITLE
Restart workers when worker-ttl expires

### LINKED ISSUE #8537: worker-ttl timeout should attempt a nanny restart
The scheduler forcefully disconnects a worker
- 30 seconds (`distributed.comm.timeouts.tcp`) after the linux kernel of the host running the worker became unresponsive, or 
- 300 seconds (`distributed.scheduler.worker-ttl`) after the worker's event loop became unresponsive,

whichever happens first.
In both cases, if the nanny is still responsive, it shuts down permanently.

In the second case, where the Worker process is KO but the underlying network and kernel are healthy, we're missing a trick where we should instead ask the nanny to restart the worker.
This is particularly important e.g. on static clusters, where there isn't any additional healing system available.

Additionally, when the worker process is so borked that it won't even notice that the batched comms have been shut down (e.g. a GIL-holding C package went into an infinite loop, or a poorly written WorkerPlugin or async task did so), the nanny won't restart it.
Again, this could be fixed by asking the nanny to restart the worker.

#### An actual example - post mortem
`@coiled.function` clusters, namely, feature 1 "static" worker that runs on the same host as the scheduler. if that worker dies, nothing will bring it back up. For example, here the worker was struggling with TCP connections, which caused the `worker-ttl` lapse. Twice.

The first time, it was resurrected by the nanny. The second time it wasn't:
image

The process:
1. when the `worker-ttl` expires, the scheduler calls `Scheduler.remove_worker(close=True)`,
2. which in turn sends a batched comms message to the worker immediately before they are shut down,
3. which, when and if it arrives, triggers `Worker.close(nanny=True)`, 
4. which triggers an RPC call from worker to the nanny,
5. which, when and if it arrives, sets the nanny to `status=closing_gracefully`,
6. which, when the worker shuts itself down eventually afterwards, prevents the nanny from restarting it.

What happened in the first case, where the nanny restarted the worker? I think it's that the RPC call from worker to nanny failed, so the worker gave up and shut itself down without informing the nanny first.

### PR DESCRIPTION
**Pull request by crusaderky**
- Closes #8537
- Blocked by #8550

### CODE REVIEW

File: distributed/scheduler.py [hunk 1]
```diff
@@ -545,7 +545,7 @@ def __init__(
         self._memory_unmanaged_old = 0
         self._memory_unmanaged_history = deque()
         self.metrics = {}
-        self.last_seen = 0
+        self.last_seen = time()
```
Comment (crusaderky) [inline distributed/scheduler.py:548]: Align behaviour to ClientState

File: distributed/scheduler.py [hunk 2]
```diff
@@ -6308,13 +6321,15 @@ async def broadcast(
...
             else:
                 workers = []
+        else:
+            workers = list(workers)
         if hosts is not None:
             for host in hosts:
-                dh: dict = self.host_info.get(host)  # type: ignore
+                dh = self.host_info.get(host)
                 if dh is not None:
                     workers.extend(dh["addresses"])
```
Comment (crusaderky) [inline distributed/scheduler.py:6330]: Fix antipattern of altering function parameters

File: distributed/scheduler.py [hunk 3]
```diff
@@ -8253,19 +8268,61 @@ async def get_worker_monitor_info(self, recent=False, starts=None):
...
-                    "Worker failed to heartbeat within %s seconds. Closing: %s",
-                    self.worker_ttl,
-                    ws,
+                    f"Worker failed to heartbeat for {last_seen} seconds; {msg}: {ws}"
                 )
-                await self.remove_worker(address=ws.address, stimulus_id=stimulus_id)
+
+        if to_restart:
+            result = await self.broadcast(
+                msg={"op": "restart", "reason": stimulus_id},
```
Comment (crusaderky) [inline distributed/scheduler.py:8300]: out of scope: I'm not a fan of the `reason` parameter; we should replace it everywhere with `stimulus_id`.

**Comment by phofl** [comment]
Will this restart the worker immediately or when the thing that is running releases the Gil again/finishes?

What I want to know: If my program holds the Gil and thus the 300s timeout hits, do we restart the worker before my program finishes?

File: distributed/scheduler.py [hunk 4]
```diff
@@ -8404,19 +8407,29 @@ async def get_worker_monitor_info(self, recent=False, starts=None):
...
-                    "Worker failed to heartbeat within %s seconds. Closing: %s",
-                    self.worker_ttl,
-                    ws,
+                    f"Worker failed to heartbeat for {last_seen:.0f}s; "
+                    f"{'attempting restart' if ws.nanny else 'removing'}: {ws}"
                 )
-                await self.remove_worker(address=ws.address, stimulus_id=stimulus_id)
+
+        if to_restart:
+            await self.restart_workers(
```
Comment (phofl) [inline distributed/scheduler.py:8428]: Could you briefly check that the tracking in https://github.com/coiled/platform/blob/4dbd6f449884464caaba09b470aa06394a22d024/analytics/preload_scripts/telemetry.py#L772 still works? I don't see a reason why not, but want to be sure

Comment (crusaderky) [reply distributed/scheduler.py:8428]: Afraid I can no longer do that. Is there anything else I can do to push this PR through the finishing line?

Comment (fjetter) [reply distributed/scheduler.py:8428]: It _should_ still work from what I can tell. Closing workers are a bit of a brittle thing, though, so it's not impossible that there is some kind of race condition ongoing where that condition would no longer work. If that's the case, we can look into it later

**Comment by fjetter** [comment]
I merged main. Assuming CI is not horribly broken, we can merge

**Pull request merged**